File: 894-all-possible-full-binary-trees/894-all-possible-full-binary-trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number} n
 * @return {TreeNode[]}
 */
var allPossibleFBT = function(n) {
  if (n % 2 === 0) return []
  if (n === 1) return [new TreeNode(0)]
  const ans = []
  for (let l = 1; l < n; l += 2) {
    allPossibleFBT(l).forEach(lNode =>
      allPossibleFBT(n-l-1).forEach(rNode =>
        ans.push(new TreeNode(0, lNode, rNode))))
  }
  return ans
};

File: 894-all-possible-full-binary-trees/894-all-possible-full-binary-trees.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func allPossibleFBT(n int) []*TreeNode {
  if n % 2 == 0 { return []*TreeNode{} }
  level := (n + 1) / 2 
  dp := make([][]*TreeNode, level)
  dp[0] = []*TreeNode{&TreeNode{0, nil, nil}}
  for i := 1; i < level; i ++ {
    var si []*TreeNode
    for l := 0; l < i; l ++ {
      for _, lNode := range dp[l] {
        for _, rNode := range dp[i-l-1] {
          si = append(si, &TreeNode{0, lNode, rNode})
        }
      }
    }
    dp[i] = si
  }
  return dp[level-1]}